src/lib.rs

```diff
@@ -361,8 +386,22 @@ impl<P: AsRef<[u8]>, T: Transitions> AcAutomaton<P, T> {
                     }
                     let ufail = self.states[v as usize].goto(c);
                     self.states[u as usize].fail = ufail;
-                    let ufail_out = self.states[ufail as usize].out.clone();
-                    self.states[u as usize].out.extend(ufail_out);
+
+                    fn get_two<T>(xs: &mut [T], i: usize, j: usize) -> (&mut T, &mut T) {
+                        if i < j {
+                            assert!(j < xs.len());
+                        } else {
+                            assert!(i != j && i < xs.len());
+                        }
+                        unsafe {
+                            let x = &mut *(xs.get_unchecked_mut(i) as *mut T);
+                            let y = &mut *xs.get_unchecked_mut(j);
+                            (x, y)
+                        }
+                    }
```

Comment:
Could you please explain why you added this? `unsafe` code should be strongly justified, and IMO, the bar should be set pretty high.

Reply:
Replaced it with a safe version using `split_at_mut` (thought that version would be messier than it was).

Reply:
The reason for `get_two` is that it makes it possible to get two indices into the same slice so that the allocation could be eliminated.